parsers/NTESMO.py

```diff
@@ -71,90 +69,91 @@ class Generator(TypedDict):
     status_forcelist=[500, 502, 503, 504],
 )
 
-_DT_CLASS = "smp-tiles-article__title"
 
+def _find_link_to_daily_report(target_datetime: datetime, session: Session) -> str:
+    """Scrapes daily report cards to find the link to the data for the target date."""
 
-def construct_latest_index(session: Session) -> dict[date, str]:
-    """Browse all links from the latest daily reports page and index them."""
-    index = {}
-    latest_index_page = session.get(LATEST_URL)
-    soup = BeautifulSoup(latest_index_page.text, "html.parser")
-    for a in soup.find_all("a", href=True):
-        if a["href"].startswith(DATA_DOC_PREFIX):
-            dt = pd.to_datetime(a.find("div", {"class": _DT_CLASS}).text)
-            index[dt.date()] = a["href"]
-    return index
+    today_date = datetime.now(AUSTRALIA_TZ).date()
+    target_datetime = target_datetime.astimezone(AUSTRALIA_TZ)
 
+    # cap target datetime if more recent that what the API makes available:
+    # data is being published after 2 days at the moment
+    _DELAY_IN_DAYS = 2
+    if abs(today_date - target_datetime.date()).days < _DELAY_IN_DAYS:
+        target_datetime = target_datetime - timedelta(days=_DELAY_IN_DAYS)
 
-def construct_year_index(year: int, session: Session) -> dict[date, str]:
-    """Browse all links on a yearly historical daily data and index them."""
-    index = {}
-    # For the current we need to go to the default page.
-    url = DEFAULT_URL
-    if year != datetime.now(tz=AUSTRALIA_TZ).year:
-        url = INDEX_URL.format(year)
-    year_index_page = session.get(url)
-    soup = BeautifulSoup(year_index_page.text, "html.parser")
-    for a in soup.find_all("a", href=True):
-        if a["href"].startswith(DATA_DOC_PREFIX):
-            dt = pd.to_datetime(a.find("div", {"class": _DT_CLASS}).text)
-            index[dt.date()] = a["href"]
-    return index
-
+    this_year = target_datetime.year == today_date.year
+    if this_year:
+        # current year's report cards are paginated (9 per page)
+        num_pages = int(math.ceil(365 / 9))
+        urls = (
+            f"{API_URL}?result_70160_result_page={page_number}"
+            for page_number in range(1, num_pages + 1)
+        )
```

Comment:
thought: we could maybe even use some logic to pre-compute which page_number the data for the requested day would fall in, and query that page directly

But maybe that would be too risky?